### PR TITLE
Dynamic environment versions support

### PR DESCRIPTION
**Pull request by La-Pulga**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -41,6 +41,7 @@ gem 'htmlbeautifier'
 group :development, :test do
   # Call 'byebug' anywhere in the code to stop execution and get a debugger console
   gem 'byebug'
+  gem 'debbie'
```
Comment (bmurzeau) [inline Gemfile]: According to its [rating](https://github.com/goodpeople/debbie) this gem seems is not widely used

File: app/controllers/instances_controller.rb [hunk 2]
```diff
@@ -1,14 +1,35 @@
...
-    response = conn.get '/versions.json'
+    response = conn.get  # Parser l'url
     response.body
   end
+
+  #Récuperer les environements et utiliser la fonction sur chaque env (parser les url pour avoir un array)
+  def index
+    envs = Environment.where.not('url' => nil)
+
+    @coconut = Hash.new #Creer une hashmap qui stockera le nom du MP associé aux versions
```
Comment (bmurzeau) [inline app/controllers/instances_controller.rb]: * avoid funking variable names

File: lib/Version.rb [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+  def call(env_url)
+    # Prendre l'url d'un environement
+    conn = Faraday.new(:url => env_url)
+
+    conn.adapter Faraday.default_adapter
+    conn.response :json
+    # Parser l'url
+    response = conn.get
+    response.body
+    binding.pry
```
Comment (ArmandPredicSis) [inline lib/Version.rb]: remove

**Comment by ArmandPredicSis** [comment]
* In this PR, you have three methods `def call` which do the same thing (or almost). Keep only the good one.

File: config/application.rb [hunk 4]
```diff
@@ -6,11 +6,11 @@
...
-module Get_Version
+module Version
   class Application < Rails::Application
-  	config.autoload_paths += %W(#{config.root}/lib) 
-    # Settings in config/environments/* take precedence over those specified here.
-    # Application configuration should go into files in config/initializers
-    # -- all .rb files in that directory are automatically loaded.
-  end
-end
+    config.autoload_paths += %W(#{config.root}/lib)
```
Comment (ArmandPredicSis) [inline config/application.rb]: Wrong indentation

**Comment by ArmandPredicSis** [comment]
* Remove comments in `app/controllers/instances_controller.rb`, `app/models/environment.rb`, `faraday.rb` & `lib/Version.rb`

File: app/controllers/instances_controller.rb [hunk 5]
```diff
@@ -1,14 +1,47 @@
...
 
-    response = conn.get '/versions.json'
+    response = conn.get  # Parser l'url
     response.body
   end
+
+  #Récuperer les environements et utiliser la fonction sur chaque env (parser les url pour avoir un array)
+  def index
+    envs = Environment.where.not('url' => nil) #Array (du modele) d'objets de type environnement.
+    @envs = envs
```
Comment (ArmandPredicSis) [inline app/controllers/instances_controller.rb]: You can directly do: `@envs = Environment.where.not('url' => nil)`

File: app/controllers/instances_controller.rb [hunk 6]
```diff
@@ -1,14 +1,47 @@
...
+
+    @app = Hash.new #Creer une hashmap qui stockera le nom des applications associées aux versions
+
+    envs.each do |e|  #Creer iteration avec un nom pour la variable locale
+      result = call(e.url)#Mettre le résultat dans une nouvelle variable. Returning parsed list of URLs =Json objects (Json objects= hashmap).
+
+      result.each do |n|
+
+        @app[n["name"]] ||= {} #Create hash if key not present. If key is present, leave hash unchanged.
+        env_to_app =  @app[n["name"]]
```
Comment (ArmandPredicSis) [inline app/controllers/instances_controller.rb]: Unecessary assignment.

File: app/controllers/instances_controller.rb [hunk 7]
```diff
@@ -1,14 +1,47 @@
...
+        version = n["version"]
+
+
+        env_to_app.merge!(e.name => version) #Adds the content of env_name to @app. @app = {app_name => env_to_app}
```
Comment (ArmandPredicSis) [inline app/controllers/instances_controller.rb]: You can directly do `(@app[n["name"]] || {}).merge!(e.name => version)`

File: app/controllers/instances_controller.rb [hunk 8]
```diff
@@ -1,14 +1,47 @@
...
+
+      end
+    end
+  end
 end
+
+
+
+
+#@app = {nom_app => {env_name , version}}
```
Comment (ArmandPredicSis) [inline app/controllers/instances_controller.rb]: Remove comments ;)

File: config/routes.rb [hunk 9]
```diff
@@ -1,5 +1,6 @@
 Rails.application.routes.draw do
-  
+
+  resources :instances
   get 'instances/index'
```
Comment (ArmandPredicSis) [inline config/routes.rb]: Can be removed.

File: config/routes.rb [hunk 10]
```diff
@@ -1,5 +1,6 @@
 Rails.application.routes.draw do
-  
+
+  resources :instances
```
Comment (ArmandPredicSis) [inline config/routes.rb]: This notation allows these actions on `instances_controller` :
- `show`
- `index`
- `new`
- `create`
- `edit`
- `update`
- `destroy`

You only want the `index` method, so youy can specify it like this: `resources :instances, only: :index`

File: config/application.rb [hunk 11]
```diff
@@ -6,11 +6,11 @@
 # you've limited to :test, :development, or :production.
 Bundler.require(*Rails.groups)
 
-module Get_Version
+module Version
```
Comment (ArmandPredicSis) [inline config/application.rb]: This module name should be `RevisionObserver` (the name of your application)

File: db/routes.rb [hunk 12]
```diff
@@ -0,0 +1,11 @@
+Rails.application.routes.draw do
```
Comment (ArmandPredicSis) [inline db/routes.rb]: Remove this module. (`routes.rb`must be in the `config` folder)

File: test/models/environment_test.rb [hunk 13]
```diff
@@ -0,0 +1,24 @@
+require 'test_helper'
+
+class EnvironmentTest < ActiveSupport::TestCase
```
Comment (bmurzeau) [inline test/models/environment_test.rb]: 👍

**Pull request merged**